# Townsend Lab
This repository contains the code and some results for some projects I have been conducting at the Townsend Lab at the Yale School of Public Health. Please check out our lab's [GitHub](https://github.com/Townsend-Lab-Yale) for more information about the research we perform!

## Melanoma Stages
This folder contains the code and results for the project described below. This project was presented at the American Association for Cancer Research's (AACR) special conference on translating cancer evolution and data science.

Project Title: Quantification of cancer effect sizes of somatic variants reveals stage-dependent evolution in primary and metastatic skin cutaneous melanoma
Description: The identities of driver and passenger genes in skin cutaneous melanoma (SKCM) have been intensively studied. However, the relative enhancement of cancer cell lineage survival and proliferation by SKCM driver mutations, as well as how these cancer effects vary with cancer stage, remains limited. The cancer effect size (also known as the scaled selection coefficient) quantifies the relative cancer cell lineage survival and proliferation in the context of each driver mutation. Using a diverse set of 999 primary and 1149 metastatic SKCM samples from multiple datasets, including The Cancer Genome Atlas (TCGA) and the AACR Project GENIE, we computed the cancer effect size of all recurrent mutations in primary (Stage Iâ€“III) and metastatic (Stage IV) tumors. Our analysis revealed that both primary and metastatic tumors were alike in their trinucleotide mutation profiles. However, the underlying gene mutation rates typically increased from primary to metastatic tumors as the latter group typically features ineffective DNA repair machinery. Additionally, the two cohorts noticeably differed in both magnitudes of effect sizes and in which genes specific somatic variants were most highly selected. Effect sizes of several driver mutations, most notably BRAF V600E, were orders of magnitude higher in metastatic tumors than in primary tumors. Furthermore, certain variants, particularly NONO R184L, were highly selected along the trajectory from normal tissue to primary tumors, yet new mutations were no longer selected between primary tumor resection and the development of resectable metastases. Conversely, for several variants, including H3F3B R18L, TP53 R273L, and PIK3CA R115P, selection was very strong only along the trajectory from resectable primary tumors to metastases. These findings illustrate the shifting evolutionary trajectory of tumors across cancer stages, underscoring the importance of early detection of SKCM and informing the development of targeted therapeutics aimed at specific stages of tumorigenesis.


## PDAC Drivers
This folder contains the code and results related to a project that seeks rank the relative contribution of somatic mutations on cancer cell lineage survival and proliferation in pancreatic ductal adenocarciomas (PDACs). We compiled data from over 6,000 PDAC exomes to perform our analysis. We explored the pairwise epistatic effects between PDAC driver genes. The ultimate goal of such analyses is to better guide the development of targeted therapeutics for various cancer types. 

